referral_module/tests/test_middleware.py

```diff
@@ -0,0 +1,96 @@
+from django.test import TestCase
+from django.http import HttpRequest
+from django.http import HttpResponseBadRequest
+
+from referral_module import middleware, models, constants
+from .factories import UserReferrerFactory, CampaignFactory
+
+
+def request_response_helper(cp, uk):
+    request = HttpRequest()
+    request.META['REMOTE_ADDR'] = '192.168.0.100'
+    data_dict = {'cp': cp, 'uk': uk}
+    request.GET = data_dict
+    referrer_store = middleware.ReferrerStoreMiddleware(
+        lambda request: HttpResponseBadRequest()
+    )
+    return referrer_store(request)
+
+
+class MiddlewareTestCase(TestCase):
+    def setUp(self):
+        campaign = CampaignFactory(
+            bonus_policy={
+                'click': 6,
+                'registration': 7,
+            }
+        )
+        self.user_referrer = UserReferrerFactory(
+            campaign=campaign
+        )
+
+        self.cp = self.user_referrer.campaign.key
+        self.uk = self.user_referrer.key
+        self.user_referrer_db = models.UserReferrer.objects.get(key=self.uk)
+
+    def test_middleware_update_referrer_data(self):
+        reward_before = self.user_referrer_db.reward
+        response = request_response_helper(cp=self.cp, uk=self.uk)
+
+        # check response status code
+        self.assertEqual(response.status_code, 400)
```

Comment:
I think in general it is a bit wired that we count rewards event if response status code is `400` which is `bad request`